cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -86,8 +88,106 @@ void CoordinateElement::compute_reference_geometry(
   assert(K.dimension(1) == this->topological_dimension());
   assert(K.dimension(2) == this->geometric_dimension());
 
-  assert(_compute_reference_geometry);
-  _compute_reference_geometry(X.data(), J.data(), detJ.data(), K.data(),
-                              num_points, x.data(), cell_geometry.data());
+  const int d = cell_geometry.rows();
+  Eigen::VectorXd phi(d);
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> dphi(
+      d, _tdim);
+
+  if (_is_affine)
+  {
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> x0(_gdim);
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> X0(_tdim);
+    X0.setZero();
+    // Compute physical coordinates at X=0.
+    _evaluate_basis_derivatives(phi.data(), 0, 1, X0.data());
+    x0 = cell_geometry.matrix().transpose() * phi;
+
+    // Compute Jacobian and inverse
+    _evaluate_basis_derivatives(dphi.data(), 1, 1, X0.data());
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor, 3, 3>
+        J0(_gdim, _tdim);
+    J0 = cell_geometry.matrix().transpose() * dphi;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor, 3, 3>
+        K0(_tdim, _gdim);
+    if (_gdim == _tdim)
+    {
+      K0 = J0.inverse();
+      detJ.fill(J0.determinant());
+    }
+    else
+    {
+      // Penrose-Moore pseudo-inverse
+      K0 = (J0.transpose() * J0).inverse() * J0.transpose();
+      detJ.fill(std::sqrt((J0.transpose() * J0).determinant()));
+    }
+
+    // Fill result for J, K and detJ
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Jview(J.data(), _gdim * num_points, _tdim);
+    Jview = J0.replicate(num_points, 1);
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Kview(K.data(), _tdim * num_points, _gdim);
+    Kview = K0.replicate(num_points, 1);
+
+    // Calculate X for each point
+    for (int ip = 0; ip < num_points; ++ip)
+      X.row(ip) = K0 * (x.row(ip).matrix().transpose() - x0);
+  }
+  else
+  {
+    // Newton's method for non-affine geometry
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> xk(x.cols(),
+                                                                       1);
+    Eigen::Matrix<double, Eigen::Dynamic, 1, Eigen::ColMajor, 3, 1> Xk(_tdim,
+                                                                       1);
+
+    for (int ip = 0; ip < num_points; ++ip)
+    {
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Jview(J.data() + ip * _gdim * _tdim, _gdim, _tdim);
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Kview(K.data() + ip * _gdim * _tdim, _tdim, _gdim);
+      // TODO: Xk - use cell midpoint instead?
+      Xk.setZero();
+      const int max_its = 10;
```

Comment:
I think this could be a kwarg now, as we are throwing an error if we hit max iteration count. Then we could adjust the error message to:  "Newton method failed to converge for non-affine geometry. Try increasing the maximum number of iterations".